nni/algorithms/compression/pytorch/quantization/quantizers.py

```diff
@@ -120,6 +123,233 @@ def quant_backward(tensor, grad_output, quant_type, scale, zero_point, qmin, qma
+            if "output" in config.get("quant_types", []):
+                all_observers[layer_name]["output"] = default_histogram_observer()
+                setattr(module, "output_qmax", activation_q_max)
+                setattr(module, "output_qmin", activation_q_min)
+        self.all_observers = all_observers
+        self.bound_model.to(self.device)
+
+    def validate_config(self, model, config_list):
+        schema = QuantizerSchema([{
+            Optional('quant_types'): Schema([lambda x: x in ['weight', 'output', 'input']]),
+            Optional('quant_bits'): Or(And(int, lambda n: n == 8), Schema({
+                Optional('weight'): And(int, lambda n: n == 8),
+                Optional('output'): And(int, lambda n: n == 8),
+                Optional('input'): And(int, lambda n: n == 8),
+            })),
+            Optional('op_types'): [str],
+            Optional('op_names'): [str]
+        }], model, logger)
+
+        schema.validate(config_list)
+
+    def record(self, wrapper, quant_type, tensor):
+        name = wrapper.name
+        observer = self.all_observers[name][quant_type]
+        if isinstance(tensor, tuple):
+            # NB: This only works for single tensor
+            tensor = (t.cpu() for t in tensor)
+            observer(*tensor)
+        else:
+            observer(tensor.cpu())
+
+    def calculate_qparams(self, name, quant_type):
+        observer = self.all_observers[name][quant_type]
+        scale, zero_point = observer.calculate_qparams()
+        return scale, zero_point
+
+    def _quantize(self, x, scale, zero_point, qmin, qmax):
+        x = x / scale + zero_point
+        x = torch.clamp(x, qmin, qmax)
+        x = torch.round(x)
+        x = (x - zero_point) * scale
+        return x
+
+    def quantize_input(self, *inputs, wrapper, **kwargs):
+        if self.compressed:
+            module = wrapper.module
+            new_input = self._quantize(inputs[0],
+                                      module.input_scale,
+                                      module.input_zero_point,
+                                      module.input_qmin,
+                                      module.input_qmax)
+            list_inp = list(inputs)
+            list_inp[0] = new_input
+            inputs = tuple(list_inp)
+        else:
+            self.record(wrapper, 'input', inputs)
+        return inputs
+
+    def quantize_weight(self, wrapper, **kwargs):
+        module = wrapper.module
+        old_weight = module.weight
+        if self.compressed:
+            new_weight = self._quantize(old_weight,
+                                       module.weight_scale,
+                                       module.weight_zero_point,
+                                       module.weight_qmin,
+                                       module.weight_qmax)
+        else:
+            self.record(wrapper, 'weight', old_weight)
+            new_weight = old_weight
+        return new_weight
```

Comment:
seems we never use this `new_weight`? because there isn't something like `module.weight = new_weight`

Reply:
Yes, `new_weight` will not be used. I just return it like what QAT quantizer does. It also returns unused new_weight. Do I need to delete it?

Reply:
I think for the quantization simulation after compress(), we need to add `module.weight = new_weight`.
```
        if self.compressed:
            new_weight = self._quantize(old_weight,
                                       module.weight_scale,
                                       module.weight_zero_point,
                                       module.weight_qmin,
                                       module.weight_qmax)
            module.weight = new_weight
```
Will we get the correct inference result in this way?

Reply:
You are right, have corrected it